paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt

```diff
@@ -51,21 +56,43 @@ import kotlin.io.path.invariantSeparatorsPathString
 class PaparazziPlugin : Plugin<Project> {
   override fun apply(project: Project) {
     project.afterEvaluate {
-      check(!project.plugins.hasPlugin("com.android.application")) {
-        error(
-          "Currently, Paparazzi only works in Android library -- not application -- modules. " +
-            "See https://github.com/cashapp/paparazzi/issues/107"
-        )
-      }
-      check(project.plugins.hasPlugin("com.android.library")) {
-        "The Android Gradle library plugin must be applied for Paparazzi to work properly."
+      val newResourceLoadingEnabled =
+        (project.findProperty("app.cash.paparazzi.new.resource.loading") as? String)?.toBoolean() ?: false
+
+      if (!newResourceLoadingEnabled) {
+        check(!project.plugins.hasPlugin("com.android.application")) {
+          error(
+            "Currently, Paparazzi only works in Android library -- not application -- modules. " +
+              "See https://github.com/cashapp/paparazzi/issues/107"
+          )
+        }
+
+        check(project.plugins.hasPlugin("com.android.library")) {
+          "The Android Gradle library plugin must be applied for Paparazzi to work properly."
+        }
+
+        project.plugins.withId("com.android.library") {
+          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+        }
+      } else {
+        check(listOf("com.android.application", "com.android.library", "com.android.dynamic-feature").any { project.plugins.hasPlugin(it) }) {
+          "The Android Gradle application or library or dynamic feature plugin must be applied for Paparazzi to work properly."
+        }
+
+        project.plugins.withId("com.android.application") {
+          setupPaparazzi(project, project.extensions.getByType(BaseAppModuleExtension::class.java).applicationVariants)
+        }
+        project.plugins.withId("com.android.dynamic-feature") {
+          setupPaparazzi(project, project.extensions.getByType(DynamicFeatureExtension::class.java).applicationVariants)
+        }
+        project.plugins.withId("com.android.library") {
+          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+        }
```

Comment:
These are deprecated APIs:
```
w: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt:22:37 'BaseVariant' is deprecated. Deprecated in Java
w: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt:23:46 'TestedVariant' is deprecated. Deprecated in Java
w: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt:95:78 'BaseVariant' is deprecated. Deprecated in Java
w: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt:111:37 'TestedVariant' is deprecated. Deprecated in Java
```
`.applicationVariants` and `.libraryVariants` are the AGP <8 APIs, since AGP 7.0 they introduced the incubating `androidComponents.beforeVariants/onVariants { }` which is now stable in 8.

Reply:
I have recently done a similar migration I'll have a look if it's easily doable here too.

Reply:
Here's a simple patch on top of [7b433ef](https://github.com/cashapp/paparazzi/pull/960/commits/7b433ef570ae7846823ae2cc5f1ab36339d049b8):

<details><summary>Click to expand</summary>

```patch
Index: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt b/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt
--- a/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt	(revision 7b433ef570ae7846823ae2cc5f1ab36339d049b8)
+++ b/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt	(revision 9da10daa1bf6f80fa098b05c633b97810dc1b7ed)
@@ -17,16 +17,11 @@
 
 import app.cash.paparazzi.gradle.utils.artifactViewFor
 import app.cash.paparazzi.gradle.utils.artifactsFor
+import com.android.build.api.variant.AndroidComponentsExtension
 import com.android.build.gradle.BaseExtension
-import com.android.build.gradle.LibraryExtension
-import com.android.build.gradle.api.BaseVariant
-import com.android.build.gradle.internal.api.TestedVariant
-import com.android.build.gradle.internal.dsl.BaseAppModuleExtension
-import com.android.build.gradle.internal.dsl.DynamicFeatureExtension
+import com.android.build.gradle.internal.component.ComponentCreationConfig
 import com.android.build.gradle.internal.publishing.AndroidArtifacts.ArtifactType
-import com.android.build.gradle.tasks.MergeSourceSetFolders
 import org.gradle.api.DefaultTask
-import org.gradle.api.DomainObjectSet
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.artifacts.component.ProjectComponentIdentifier
@@ -40,7 +35,6 @@
 import org.gradle.api.plugins.JavaBasePlugin
 import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.PathSensitivity
-import org.gradle.api.tasks.TaskProvider
 import org.gradle.api.tasks.options.Option
 import org.gradle.api.tasks.testing.Test
 import org.gradle.internal.os.OperatingSystem
@@ -72,7 +66,7 @@
         }
 
         project.plugins.withId("com.android.library") {
-          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+          setupPaparazzi(project)
         }
       } else {
         check(listOf("com.android.application", "com.android.library", "com.android.dynamic-feature").any { project.plugins.hasPlugin(it) }) {
@@ -80,19 +74,19 @@
         }
 
         project.plugins.withId("com.android.application") {
-          setupPaparazzi(project, project.extensions.getByType(BaseAppModuleExtension::class.java).applicationVariants)
+          setupPaparazzi(project)
         }
         project.plugins.withId("com.android.dynamic-feature") {
-          setupPaparazzi(project, project.extensions.getByType(DynamicFeatureExtension::class.java).applicationVariants)
+          setupPaparazzi(project)
         }
         project.plugins.withId("com.android.library") {
-          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+          setupPaparazzi(project)
         }
       }
     }
   }
 
-  private fun setupPaparazzi(project: Project, variants: DomainObjectSet<out BaseVariant>) {
+  private fun setupPaparazzi(project: Project) {
     project.addTestDependency()
     val nativePlatformFileCollection = project.setupNativePlatformDependency()
 
@@ -106,14 +100,14 @@
       it.description = "Record golden images for all variants"
     }
 
-    variants.all { variant ->
+    val androidComponents = project.extensions.getByType(AndroidComponentsExtension::class.java)
+    androidComponents.onVariants { variant ->
       val variantSlug = variant.name.capitalize(Locale.US)
-      val testVariant = (variant as TestedVariant).unitTestVariant ?: return@all
+      val testVariant = variant.unitTest ?: return@onVariants
 
-      val mergeResourcesOutputDir = variant.mergeResourcesProvider.flatMap { it.outputDir }
-      val mergeAssetsProvider =
-        project.tasks.named("merge${variantSlug}Assets") as TaskProvider<MergeSourceSetFolders>
-      val mergeAssetsOutputDir = mergeAssetsProvider.flatMap { it.outputDir }
+      val taskContainer = (variant as ComponentCreationConfig).taskContainer // Internal API!
+      val mergeResourcesOutputDir = taskContainer.mergeResourcesTask.flatMap { it.outputDir }
+      val mergeAssetsOutputDir = taskContainer.mergeAssetsTask.flatMap { it.outputDir }
       val projectDirectory = project.layout.projectDirectory
       val buildDirectory = project.layout.buildDirectory
       val gradleUserHomeDir = project.gradle.gradleUserHomeDir
@@ -121,7 +115,7 @@
       val snapshotOutputDir = project.layout.projectDirectory.dir("src/test/snapshots")
 
       val localResourceDirs = project
-        .files(variant.sourceSets.flatMap { it.resDirectories })
+        .files(variant.sources.res!!.all) // TODO Review if this is what you need.
 
       // https://android.googlesource.com/platform/tools/base/+/96015063acd3455a76cdf1cc71b23b0828c0907f/build-system/gradle-core/src/main/java/com/android/build/gradle/tasks/MergeResources.kt#875
 
```
</details>

it uses roughly the same level of hackery (internal APIs / assumptions), but no deprecated APIs.

_Note: this is fully untested, I was happy to get to a compiling stage :)_

<details><summary>A warning about using TaskProviders instead of Artifacts</summary>

Context is how we acquire `mergeResourcesOutputDir` and `mergeAssetsOutputDir` in the above code.

Using `tasks.named(guess based on variant slug)` is untyped and based on a lot of assumptions.
Using `TaskContainer` is typed, but uses an Internal API, and there's a huge warning on top of the class:
> DO NOT USE THESE GETTERS FOR TASK CONFIGURATION, USE ARTIFACTS INSTEAD

Based on this and other news, in the future they'll migrate everything to artifacts and probably remove all task getters.
For this reason I would recommend this approach:
```kotlin
val internalArtifacts = variant.artifacts as ArtifactsImpl
val mergeResourcesOutputDir = internalArtifacts.get(InternalArtifactType.MERGED_RES)
val mergeAssetsOutputDir = internalArtifacts.get(InternalArtifactType.LIBRARY_ASSETS) // or variant.artifacts.get(SingleArtifact.ASSETS)
```
it still uses an internal API, but in case they expose those `InternalArtifactType`s as public, the migration is trivial.
Also, it doesn't need to reference the task name, task container, task type and task properties, which is a lot of assumptions.
The reason why AGP introduced the artifact abstraction is to move away from *how* the files are juggle around, and focus on *what* files are needed.

Here's an annotated version of the same, detailing how I got to those conclusions:
```kotlin
// Need to cast to ArtifactsImpl from Artifacts,
// so we gain access to the .get(Single<>) overload which is the base type for both SingleArtifact and InternalArtifactType.
val internalArtifacts = variant.artifacts as ArtifactsImpl
// See com.android.build.gradle.tasks.MergeResources.CreationAction.handleProvider
// Which wires MergeResources.outputDir to mergeType, defined at com.android.build.gradle.internal.TaskManager.MergeType.
// This is the same task as TaskContainer.mergeResourcesTask which was exposed as BaseVariantImpl.mergeResourcesProvider.
val mergeResourcesOutputDir = internalArtifacts.get(InternalArtifactType.MERGED_RES)
// Instead of hardcoding task name and type: task(merge${variantSlug}Assets) as MergeSourceSetFolders
// See com.android.build.gradle.tasks.MergeSourceSetFolders.LibraryAssetCreationAction which wires this artifact.
// Note the app version of this: com.android.build.gradle.tasks.MergeSourceSetFolders.MergeAppAssetCreationAction
// is actually exposed as public API: variant.artifacts.get(SingleArtifact.ASSETS) (note: no casts needed)
val mergeAssetsOutputDir = internalArtifacts.get(InternalArtifactType.LIBRARY_ASSETS)
```
</details>

Reply:
Discussed offline, given that:
* we'll be bridging for a short period of time (1-2 months 🤞), between the old and new resource loading mechanisms  * the new APIs don't easily lend access to the internal AGP tasks (kinda the point)
* according to Xav from Google Tools team, the artifacts APIs don't currently support merged resources/assets properly (and even if they do, they can change across versions since implementation detail)

I suggest we keep the current hacky APIs until new resource loading is declared stable, then rip all the old stuff out and migrate to the new APIs which won't need the resources stuff.

I'm aware that the assets API will require a similar update; this is tracked in https://github.com/cashapp/paparazzi/issues/983.  Would be nice to tackle this for 1.3.1, but can be punted to 1.4 if the work is harder than it seems.